### DISCUSSION TITLE
Future support for Apple Silicon M1?

### DISCUSSION THREAD

**Discussion by gy2256** [cyberbotics/webots#2574] [General]
Greeting!

I am a new user of Webots and I enjoy using it! The latest Mac series have switched from the traditional x86 architecture to their ARM architecture. The performance is quite promising and many software gradually have native support [see here](https://doesitarm.com)
I am wondering if there are any plans to have future support on the Macs with ARM CPU?

I am not sure if all the dependencies support apple silicon, but at least openCL and openGL should work on the ARM Macs ([source](https://twitter.com/colincornaby/status/1275153748348682240?s=20))

**Comment by omichel** [comment]
In theory Webots R2021a should already work with Rosetta 2, but I didn't test it because I don't have a M1 computer here. Can you test it?

Of course, we have plans to support Apple M1. I believe we simply need to modify the build process (Makefiles) in order to create fat binaries that include binary code for both ARM and X86-64 architectures. It shouldn't be very difficult. If you have a M1 machine at hand, maybe you can give a try, [compiling Webots from the source](https://github.com/cyberbotics/webots/wiki/macOS-installation), [modifying the Makefiles to include ARM support](https://developer.apple.com/documentation/xcode/building_a_universal_macos_binary) and testing if it works (and if it runs faster)?

**Comment by gy2256** [comment]
Yeah, I am really interested in this too. Unfortunately, I do not have a M1 machine as well. I will definitely test it once I have the machine. 

My only concern is the lack of supports from other open-source projects. For example, ROS2 will move Mac OS from Tier 1 to Tier 2 or Tier 3 support now. In addition, there are still no native Fortran compiler or GCC compiler yet. Hopefully things will improve over the time. 

The only good news are homebrew is working hard to get things working on the ARM Mac and there is a special version of TensorFlow that utilizes the latest Neural Engine from ARM Macs [https://github.com/apple/tensorflow_macos](https://github.com/apple/tensorflow_macos)

**Comment by gy2256** [comment]
I have the M1 MacBookr Air now.

**Comment by omichel** [comment]
One major roadblock for porting Webots to M1 is Qt.
Qt doesn't yet support the M1 processor natively, but there is some [work in progress](https://bugreports.qt.io/browse/QTBUG-85279) about it and it should be available in Qt 6.2.
Webots is currently linked against Qt 5.15 on macOS, but it should be relatively easy to upgrade to Qt 6.2 when it's out.
As soon as Qt 6.2 is released with M1 support, we are going to try to build a first beta version of Webots with native M1 support.
We will post the news here with a link to the beta version if any for those interested to test it.

**Reply by gy2256** [reply]
Yeah. I have encountered the same problem (no qt) when i try to compile another program.

**Comment by Suke0811** [comment]
Is there any update on M1 chip support? 
When we try to run through 'extern' controller, however, we got an error saying architecture is incompatible as shown below.
![image](https://user-images.githubusercontent.com/49264928/151463426-8e791f03-c27a-46d1-ba05-ab617de2e223.png)

This is with M1 Macbook air, using Python 39.

**Comment by omichel** [comment]
It is possible to run Webots python controllers on a M1 Mac as explained in https://github.com/cyberbotics/webots/issues/4112#issuecomment-1012324873.

**Comment by omichel** [comment]
Native support for Apple M1 is coming now in #4543 starting with robot controllers.

**Comment by sanieakhtar** [comment]
Hi! I'm also interested in this topic. I have an M1 Macbook Air 2020. It's the "latest" Macbook Air as the new one might be released this year. I just bought my mac last month and used webots for the first time this weekend. The app crashed so many times, I am actually getting concerned about my machine.

Is it .. normal for webots to be doing this as full support is still in the works?

I was keeping an eye on the activity monitor and didn't seem like the memory was overloaded. I updated to Rosetta 2 and I'm using R2021. However, the amount of times I had to restart the webots app, for a relatively quite simple world and accompanying controller, has me concerned.

Looking forward to hearing other people's opinions..

**Reply by omichel** [reply]
We should have a native M1 version of Webots very soon. This might help resolve the crashes you experienced. Would you like to beta test this new native M1 version?

**Reply by RenzoDaal** [reply]
Yes please. I have the same issue.

**Comment by omichel** [comment]
#4543 was just merged into the develop branch. This means that from tomorrow morning (31-5-22), a universal binary (including native M1 support) distribution of Webots R2022b should be available in the [nightly builds](https://github.com/cyberbotics/webots/releases). Could you please test it and report any problem with it?

**Reply by RenzoDaal** [reply]
I have tried it and my initial problem was fixed. Rotating my view is a little different / worse. In the stable release you would rotate around the center of your world but it now seems like I am rotating around something much further away. So my objects move very slowly and I have to drag my view back down after every time I rotate.

**Reply by omichel** [reply]
You can't have tried it because it is not yet released. Check the date of the next nightly build, it should be 31-5-22.

**Comment by omichel** [comment]
The first nightly build with M1 native support is available [here](https://github.com/cyberbotics/webots/releases/download/nightly_31_5_2022/webots-R2022b.dmg). Upcoming [nightly builds](https://github.com/cyberbotics/webots/releases/) of Webots R2022b should also include this feature. Could you please test it and report any problem here?

**Reply by gy2256** [reply]
Thanks for letting us know! I am currently traveling and will test it on my MacBook Pro 14 once I get back.

**Reply by omichel** [reply]
Excellent. Let us know you when you test it.

**Reply by gy2256** [reply]
I asked a friend to try the R2022b on a MacBook Air with M1. It's amazing how well the simulator ran on the fanless laptop: the surface temperature was cool and most simple projects ran flawlessly. The setting shows the backend is OpenGL, not Metal. 

The graphics setting was

Ambient Occlusion: Ultra
Texture Quality: High
Max Texture Filtering: 5
Shadows: Enabled
Anti-aliasing: Enabled

There was an warning message in universal_robots/worlds/ure.wbt after running the simulation for a while. It shows: 

```WARNING: The current physics step could not be computed correctly. Your world may be too complex. ```

Larger project, such as vehicles/worlds/city_traffic.wbt also works but require to setup Python Path manually in Preference. 

Lastly, the extremally large project, such as vehicles/worlds/village_center.wbt was having trouble to render trees and buildings correctly.

**Reply by omichel** [reply]
Great ! Thank you for this feedback. It looks like the M1 support is working pretty nicely.

**Reply by gy2256** [reply]
No problem. I am glad that Webots is providing the Apple Silicon support. 

How difficult would it be to use the Metal backend? or is it worth the effort?

**Reply by omichel** [reply]
> How difficult would it be to use the Metal backend? or is it worth the effort?

It's a matter of porting the WREN (Webots Rendering EngiNe) library from OpenGL to Metal. This is probably a lot of work and not a priority for us now.

**Reply by gy2256** [reply]
It makes sense. As always, excellent work!

**Comment by janblumenkamp** [comment]
I just tried running the current nighly on my Macbook Air M2. If I run it, it just says that it has detected that my GPU is not from NVidia but from Apple, which makes sense. Webots starts, but the world is just black (so it seems like rendering does not work). Are there any settings I have to enable?
Thanks for clarifying,
Jan

**Reply by janblumenkamp** [reply]
Nevermind, it works! I didn't actually open my world file.

**Comment by amalalreshi** [comment]
i follow this video but i get 2d shape why ( mac m1 Os 15.1)
https://www.youtube.com/watch?v=luyg3plGujg&list=PLbEU0vp_OQkUwANRMUOM00SXybYQ4TXNF

![Screenshot 2024-11-13 at 11 07 18 PM](https://github.com/user-attachments/assets/a79831ce-2a61-4eb4-a3c5-adc20846ef04)
